R/slide.R

```diff
@@ -223,6 +221,35 @@ epi_slide = function(x, f, ..., before, after, ref_time_values,
     after <- time_step(after)
   }
 
+  inrange_time_values_not_in_x = ref_time_values - before
+  inrange_time_values_not_in_x <- inrange_time_values_not_in_x[!(inrange_time_values_not_in_x %in% unique(x$time_value))]
+
+  # Do set up to let us recover `ref_time_value`s later.
+  # A helper column marking real observations.
+  x$.real = TRUE
+
+  # Create df containing phony data. Df has the same columns and attributes as
+  # `x`, but filled with `NA`s aside from grouping columns. Number of rows is
+  # equal to the number of `inrange_time_values_not_in_x` we have * the
+  # number of unique levels seen in the grouping columns.
+  before_time_values_df = data.frame(time_value=inrange_time_values_not_in_x)
+  if (length(group_vars(x)) != 0) {
+    before_time_values_df = dplyr::cross_join(
+      # Get unique combinations of grouping columns seen in real data.
+      unique(x[, group_vars(x)]),
```

Comment:
Tried to test this here:
```{r}
as_epi_df(tibble(geo_value=c("a","a","a","b"), time_value = c(1:3, 2))) %>%
  group_by(geo_value) %>%
  epi_slide(function(x,g,t) if(nrow(x) == 0L) stop("0 row") else 1, before = 0, after = 0)
```
and got no error, so maybe this isn't happening.  It looks like the first part of `slide_one_grp` is already taking care of this.

Reply:
Great question about potential 0-row inputs! I don't see a way around the `cross_join`, but maybe we could add a short-circuit condition inside the [`f` wrapper functions](https://github.com/cmu-delphi/epiprocess/blob/0b48ea4a4ff18d9da2cade6470b424b1275be06e/R/slide.R#L357-L362). Like
```
if (all(!x$.real)) { return() }
```

It could be worth adding a test for this, too. If we change upstream behavior, we don't want to start processing 0-row inputs when we weren't before.